docs: Update Dagger API pages for Dagger shell (#9811)
dagger/dagger: docs/current_docs/api/clients-cli.mdx

---
slug: /api/cli
---
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

# Dagger CLI

The Dagger CLI lets you call both the core and extended Dagger API (the core APIs plus the new APIs provided by Dagger Functions) directly from the command-line via the `dagger shell`, `dagger core` and `dagger call` commands.

## Core Dagger API functions

The `dagger shell` and `dagger core` commands let you call the core Dagger API, without needing to first load a Dagger [module](../features/modules.mdx). This is useful in some scenarios:

- You want to use Dagger for simple pipelines that are fully satisfied by the core Dagger API.

    <Tabs>

>     <TabItem value="dagger shell">

Reviewer: ```suggestion
    <TabItem value="dagger core">
```

You can call the API interactively (`dagger`) or non-interactively (`dagger -c`, `dagger call`, or `dagger core`).

Here are a few examples:

- Create a simple pipeline that is fully satisfied by the core Dagger API, without needing to program a Dagger module:

    <Tabs groupId="shell">
    <TabItem value="System shell">
    ```shell
    dagger <<EOF
    container |
      from cgr.dev/chainguard/wolfi-base |
      with-exec apk add go |
      with-directory /src https://github.com/golang/example#master |
      with-workdir /src/hello |
      with-exec -- go build -o hello . |
      file ./hello |
      export ./hello-from-dagger
    EOF
    ```
    </TabItem>
    <TabItem value="Dagger Shell">
    ```shell title="Type 'dagger' to start Dagger Shell in interactive mode and enter the command below."
    container \
      | from cgr.dev/chainguard/wolfi-base \
      | with-exec apk add go \
      | with-directory /src https://github.com/golang/example#master \
      | with-workdir /src/hello \
      | with-exec -- go build -o hello . \
      | file ./hello \
      | export ./hello-from-dagger
    ```
    </TabItem>

>     <TabItem value="Dagger CLI">

Reviewer: Technically the Dagger Shell is a part of the Dagger CLI.

Reviewer: This was answered in https://github.com/dagger/dagger/pull/9905#discussion_r2005008758. Let's move on and update these everywhere when we have a better fit.

      file --path="./hello" \
      export --path="./hello-from-dagger"
    ```
    </TabItem>

>     <TabItem value="dagger core">

Reviewer: ```suggestion
    <TabItem value="dagger shell">
```

    ```shell

>     dagger shell -c 'container \
>       | from cgr.dev/chainguard/wolfi-base \
>       | with-exec apk,add, go \
>       | with-directory /src https://github.com/golang/example#master \
>       | with-workdir /src/hello \
>       | with-exec go,build,-o,hello,. \
>       | file ./hello \
>       | export ./hello-from-dagger'

Reviewer: ```suggestion
    dagger <<EOF
    container \
      | from cgr.dev/chainguard/wolfi-base \
      | with-exec apk add go \
      | with-directory /src https://github.com/golang/example#master \
      | with-workdir /src/hello \
      | with-exec -- go build -o hello . \
      | file ./hello \
      | export ./hello-from-dagger
    EOF
```

    ```
    </TabItem>
    </Tabs>

- You want to experiment with Dagger without having to program a Dagger module.

    <Tabs>
    <TabItem value="dagger shell">
    ```shell
    dagger shell -c 'container | from alpine | file /etc/os-release | contents'
    ```
    </TabItem>
    <TabItem value="dagger core">
    ```shell

>     dagger core container from --address=alpine file --path=/etc/os-release contents

Reviewer: Just fyi, the equivalent to `dagger core` is `dagger -n` (`-n` means `--no-mod`, while `-m` is for `--mod`).

By default `dagger` will try to load a module if you're in one, and fallback to just core. And even if a module is loaded, the core functions are also there. So what you have is correct, but if you know you only want core, adding `-n` will be faster because it skips the checking and potential loading of a module.

Reviewer 2: https://github.com/dagger/dagger/issues/9854#issuecomment-2727691596

Author: We have 24 instances of `dagger core` in our docs (not counting the changes in this PR). 

Should we change all instances of `dagger core` to `dagger -n` and remove `dagger core` from our docs?

    ```
    </TabItem>
    </Tabs>

    :::tip
    When using the Dagger CLI, all names (functions, arguments, struct fields, etc) are converted into a shell-friendly "kebab-case" style.
    :::

- Modules don't need to be installed locally. Dagger lets you consume modules from GitHub repositories and call their Dagger Functions as though you were calling them locally:

    <Tabs>
    <TabItem value="System shell">
    ```shell
    dagger <<EOF
    github.com/jpadams/daggerverse/trivy@v0.5.0 |
      scan-image ubuntu:latest
    EOF
    ```
    </TabItem>
    <TabItem value="Dagger Shell">
    ```shell title="Type 'dagger' to start Dagger Shell in interactive mode and enter the command below."
    github.com/jpadams/daggerverse/trivy@v0.5.0 | scan-image ubuntu:latest
    ```
    </TabItem>
    <TabItem value="Dagger CLI">
    ```shell
    dagger -m github.com/jpadams/daggerverse/trivy@v0.5.0 call \
      scan-image --image-ref=ubuntu:latest
    ```
    </TabItem>
    </Tabs>

- List all the Dagger Functions available in a module using context-sensitive help:

    <Tabs>
    <TabItem value="System shell">
    ```shell

>     dagger -c 'github.com/jpadams/daggerverse/trivy@v0.5.0 | .help'

Reviewer: For documenting a module `.help <module>` should be used. However it also includes constructor/entrypoint usage, so not quite the same as `dagger functions`.

Also, if the module does have required arguments just `<module> | .help` will return an error because you need to actually call the constructor in order to get its return with `<module> | .help` like this.

While you generally should do `<function> | .help` to get available functions after calling another function, for a module I suggest preferring to start with `.help <module>` instead.

Author: I tried `.help github.com/jpadams/daggerverse/trivy@v0.5.0` as you suggested but it doesn't produce the list of available functions (which is the task I'm trying to solve for in this bullet point).

Reviewer: It's in a fix in main. Will be released in v0.17. If you want to try it you need a dev build.
